flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java

```diff
@@ -0,0 +1,157 @@
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.file.PartitionFileWriter;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.file.SegmentNettyPayload;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.file.SubpartitionNettyPayload;
+import org.apache.flink.util.concurrent.FutureUtils;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * The {@link DiskCacheManager} is responsible for managing cached buffers before flushing to files.
+ */
+public class DiskCacheManager {
+
+    private final int numSubpartitions;
+
+    private final PartitionFileWriter partitionFileWriter;
+
+    private final SubpartitionDiskCacheManager[] subpartitionCacheManagers;
+
+    private CompletableFuture<Void> hasFlushCompleted;
+
+    public DiskCacheManager(
+            int numSubpartitions,
+            TieredStorageMemoryManager storageMemoryManager,
+            PartitionFileWriter partitionFileWriter) {
+        this.numSubpartitions = numSubpartitions;
+        this.partitionFileWriter = partitionFileWriter;
+        this.subpartitionCacheManagers = new SubpartitionDiskCacheManager[numSubpartitions];
+        this.hasFlushCompleted = FutureUtils.completedVoidFuture();
+
+        for (int subpartitionId = 0; subpartitionId < numSubpartitions; ++subpartitionId) {
+            subpartitionCacheManagers[subpartitionId] =
+                    new SubpartitionDiskCacheManager(subpartitionId);
+        }
+        storageMemoryManager.listenBufferReclaimRequest(this::notifyFlushCachedBuffers);
+    }
+
+    /**
+     * Append the end-of-segment event to {@link DiskCacheManager}.
+     *
+     * @param record the end-of-segment event
+     * @param subpartitionId target subpartition of this record.
+     * @param dataType the type of this record. In other words, is it data or event.
+     */
+    public void appendEndOfSegmentEvent(
+            ByteBuffer record, int subpartitionId, Buffer.DataType dataType) {
+        subpartitionCacheManagers[subpartitionId].appendEndOfSegmentEvent(record, dataType);
+        flushAndReleaseCacheBuffers();
+    }
+
+    /**
+     * Append buffer to {@link DiskCacheManager}.
+     *
+     * @param buffer to be managed by this class.
+     * @param subpartitionId the subpartition of this record.
+     */
+    public void append(Buffer buffer, int subpartitionId) {
+        subpartitionCacheManagers[subpartitionId].append(buffer);
+    }
+
+    /**
+     * Return the finished buffer index.
+     *
+     * @param subpartitionId the target subpartition id
+     * @return the finished buffer index
+     */
+    public int getFinishedBufferIndex(int subpartitionId) {
+        return subpartitionCacheManagers[subpartitionId].getFinishedBufferIndex();
```

Comment:
Which buffer is the finished buffer?

Reply:
I have removed the concept of `finished`. Rename it to `BufferIndex` directly.

Each time a new buffer is added, the field of `BufferIndex` will increase by 1.